DataStructures/Lists/CircularBuffer.cs

```diff
@@ -0,0 +1,197 @@
+            if (_end == _start) 
+            {
+                _start = (_start + 1) % _circularBuffer.Length;
+            }
+        }
+
+        /// <summary>
+        ///     Reads and removes the value in front of the buffer, and places the next value in front.
+        /// </summary>
+        public T Pop() 
+        {
+            var result = _circularBuffer[_start];
+            _circularBuffer[_start] = _circularBuffer[_end];
+            _start = (_start + 1) % _circularBuffer.Length;
+            return result;
+        }
+
+        #region IEnumerable Implementation
+        public IEnumerator<T> GetEnumerator() 
+        {
+            for (int i = 0; i < Count; i++) {
+                yield return _circularBuffer[i];
+            }
+        }
+
+        IEnumerator IEnumerable.GetEnumerator() 
+        {
+            return GetEnumerator();
+        }
+        #endregion
+
+        #region ICollection Implementation
+        /// <summary>
+        /// Returns the number of elements.
+        /// </summary>
+        public int Count 
+        {
+            get { return _length; }
+        }
+        /// <summary>
+        /// Checks whether this collection is readonly
+        /// </summary>
+        public bool IsReadOnly 
+        {
+            get { return false; }
+        }
+        /// <summary>
+        /// Clears this instance
+        /// </summary>
+        public void Clear() 
+        {
+            for (int i = 0; i < _circularBuffer.Length; i++) 
+            {
+                _circularBuffer[i] = _default;
+            }
+        }
+        /// <summary>
+        /// Checks whether the buffer contains an item
+        /// </summary>
+        public bool Contains(T item) 
+        {
+            return _circularBuffer.Contains(item);
+        }
+        /// <summary>
+        /// Copies this buffer to an array
+        /// </summary>
+        public void CopyTo(T[] array, int arrayIndex) 
+        {
+            if (array == null) 
+            {
+                throw new ArgumentNullException("array can not be null");
+            }
+
+            if (array.Length == 0 || arrayIndex >= array.Length || arrayIndex < 0) {
+                throw new IndexOutOfRangeException();
+            }
+     
+            // Get enumerator
+            var enumarator = GetEnumerator();
+
+            // Copy elements if there is any in the buffer and if the index is within the valid range
+            while (arrayIndex < array.Length) 
+            {
+                if (enumarator.MoveNext()) 
+                {
+                    array[arrayIndex] = enumarator.Current;
+                    arrayIndex++;
+                }
+                else 
+                {
+                    break;
+                }
+            }
+        }
+        /// <summary>
+        /// Removes an item from the buffer
+        /// </summary>
+        public bool Remove(T item) 
+        {
+            if (Contains(item)) 
```

Comment:
No need for the `Contains` method call, you are already looping through the collection, might as well handle the boolean check there. Also, `Contains` has an implementation of `O(n)`, which means with this implementation you're looping over the collection twice.

Reply:
Please review how it is now.

Reply:
Looks better.